transports/webrtc/src/transport.rs

```diff
@@ -0,0 +1,697 @@
+                                let ma = socketaddr_to_multiaddr(&socket_addr);
+                                debug!("New listen address: {}", ma);
+                                Some(TransportEvent::NewAddress {
+                                    listener_id: self.listener_id,
+                                    listen_addr: ma,
+                                })
+                            } else {
+                                self.poll_if_addr(cx)
+                            }
+                        }
+                        Ok(IfEvent::Down(inet)) => {
+                            let ip = inet.addr();
+                            if self.listen_addr.is_ipv4() == ip.is_ipv4()
+                                || self.listen_addr.is_ipv6() == ip.is_ipv6()
+                            {
+                                let socket_addr = SocketAddr::new(ip, self.listen_addr.port());
+                                let ma = socketaddr_to_multiaddr(&socket_addr);
+                                debug!("Expired listen address: {}", ma);
+                                Some(TransportEvent::AddressExpired {
+                                    listener_id: self.listener_id,
+                                    listen_addr: ma,
+                                })
+                            } else {
+                                self.poll_if_addr(cx)
+                            }
+                        }
+                        Err(err) => {
+                            debug! {
+                                "Failure polling interfaces: {:?}.",
+                                err
+                            };
+                            Some(TransportEvent::ListenerError {
+                                listener_id: self.listener_id,
+                                error: err.into(),
+                            })
+                        }
+                    }
+                } else {
+                    self.poll_if_addr(cx)
+                }
+            }
+            Poll::Pending => None,
+        }
+    }
+}
+
+impl Stream for WebRTCListenStream {
+    type Item = TransportEvent<<WebRTCTransport as Transport>::ListenerUpgrade, Error>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        if let Some(closed) = self.report_closed.as_mut() {
+            // Listener was closed.
+            // Report the transport event if there is one. On the next iteration, return
+            // `Poll::Ready(None)` to terminate the stream.
+            return Poll::Ready(closed.take());
+        }
+        if let Some(event) = self.poll_if_addr(cx) {
+            return Poll::Ready(Some(event));
+        }
+
+        let udp_mux = self.udp_mux.clone();
+        let udp_mux_read_fut = self
+            .udp_mux_read_fut
+            .get_or_insert(Box::pin(async move { udp_mux.read_from_conn().await }));
```

Comment:
In support of @thomaseizinger suggestion above, adding a wrapper around webrtc-rs that exposes _poll style_ methods instead of `async` methods.

In my eyes using `async` (e.g. `udp_mux.read_from_conn`) in _poll style_ methods (e.g. `WebRTCListenStream::poll_next`) adds a lot of complexity to an already complex topic, i.e. concurrency, by e.g. requiring keeping around a `udp_mux_read_fut` which in turn requires a `Mutex` to have multiple shared references to the underlying object.

I am very much in favor of using `async` in **sequential** asynchronous operations (e.g. read this, THEN write that, THEN send this, ...), but for **concurrent** asynchronous operations (e.g. try to read AND write AND also send) I find _poll style_ methods easier to reason about.

I know this is subjective. I am curious what you @melekes think, given that you are very familiar with the webrtc-rs codebase and thus the many design decisions.

Reply:
> In support of @thomaseizinger suggestion above, adding a wrapper around webrtc-rs that exposes poll style methods instead of async methods.

Do you maybe have a blog post or a reference implementation, which clearly explains the difference between those two approaches?

Reply:
> I know this is subjective. I am curious what you @melekes think, given that you are very familiar with the webrtc-rs codebase and thus the many design decisions.

There's a big discussion in webrtc-rs re/ current API and design choices and I myself not happy about some of methods and API, but it will take a long time to change it since it's a big library.

Reply:
fyi: I've just remembered. I've tried storing a future (`read_from_conn`) inside a UDP mux, but wasn't compiling because `Arc<UDPMuxNewAddr>` needs a future to be `Sync` and it's not.

Reply:
> > I know this is subjective. I am curious what you @melekes think, given that you are very familiar with the webrtc-rs codebase and thus the many design decisions.
> 
> There's a big discussion in webrtc-rs re/ current API and design choices and I myself not happy about some of methods and API, but it will take a long time to change it since it's a big library.

Oh can you post a link? I'd love to subscribe and follow the progress.

Reply:
> > In support of @thomaseizinger suggestion above, adding a wrapper around webrtc-rs that exposes poll style methods instead of async methods.
> 
> Do you maybe have a blog post or a reference implementation, which clearly explains the difference between those two approaches?

In general, the idea is:

- Have some internal state. Can be enum or a regular struct.
- Have "command"-like functions that either change state or record what needs to be done next. You can think of that as "sending a message to an actor" and "recording what needs to be done next" is basically queuing the message in the mailbox, conceptually.
- Have a `poll` function that makes progress by working off the "mailbox" and also polling inner state machines.

This design allows you to easily compose functionality on top of each other because each state machine can call `poll` on the inner one and also "send messages" to it.

Most of `rust-libp2p` would be a good reference. `libp2p-mplex` for example. The `Connection` in `libp2p-swarm` is another one that I think shows the composition aspect well.

Reply:
> > > I know this is subjective. I am curious what you @melekes think, given that you are very familiar with the webrtc-rs codebase and thus the many design decisions.
> > 
> > 
> > There's a big discussion in webrtc-rs re/ current API and design choices and I myself not happy about some of methods and API, but it will take a long time to change it since it's a big library.
> 
> Oh can you post a link? I'd love to subscribe and follow the progress.

It's on [discord](https://discord.gg/4Ju8UHdXMs) right now, but I'll post a link once there's an issue.